libsmtutil/SMTLib2Context.cpp

```diff
@@ -0,0 +1,316 @@
+}
+
+SortPointer SMTLib2Context::getDeclaredSort(std::string const& _name) const
+{
+	smtAssert(isDeclared(_name));
+	return m_functions.at(_name);
+}
+
+void SMTLib2Context::clear() {
+	m_functions.clear();
+	m_knownTypes.clear();
+	m_arraySorts.clear();
+	m_tupleSorts.clear();
+	m_bitVectorSorts.clear();
+	m_callback = {};
+	m_knownTypes.emplace_back(std::make_unique<SMTLibSort>(Kind::Bool, std::string("Bool"), std::vector<SortId>{}, SortId{0u}));
+	m_knownTypes.emplace_back(std::make_unique<SMTLibSort>(Kind::Int, std::string("Int"), std::vector<SortId>{}, SortId{1u}));
+	assert(m_boolSort.id == m_knownTypes[0]->id.id);
+	assert(m_intSort.id == m_knownTypes[1]->id.id);
+}
+
+SortId SMTLib2Context::resolve(SortPointer const& _sort)
+{
+	switch (_sort->kind)
+	{
+	case Kind::Int:
+		return m_intSort;
+	case Kind::Bool:
+		return m_boolSort;
+	case Kind::BitVector:
+		return resolveBitVectorSort(dynamic_cast<BitVectorSort const&>(*_sort));
+	case Kind::Array:
+		return resolveArraySort(dynamic_cast<ArraySort const&>(*_sort));
+	case Kind::Tuple:
+		return resolveTupleSort(dynamic_cast<TupleSort const&>(*_sort));
+	default:
+		smtAssert(false, "Invalid SMT sort");
+	}
+}
+
+SortPointer SMTLib2Context::unresolve(SortId _sort) const
+{
+	smtAssert(_sort.id < m_knownTypes.size());
+	auto const& type = *m_knownTypes[_sort.id];
+	switch (type.kind)
+	{
+	case Kind::Int:
+		return SortProvider::sintSort;
+	case Kind::Bool:
+		return SortProvider::boolSort;
+	case Kind::BitVector:
+	{
+		auto it = ranges::find_if(m_bitVectorSorts, [&](auto const& entry) { return entry.second == _sort;});
+		smtAssert(it != m_bitVectorSorts.end());
+		return std::make_shared<BitVectorSort>(it->first);
+	}
+	case Kind::Array:
+	{
+		auto it = ranges::find_if(m_arraySorts, [&](auto const& entry) { return entry.second == _sort; });
+		smtAssert(it != m_arraySorts.end());
+		return std::make_shared<ArraySort>(unresolve(it->first.first), unresolve(it->first.second));
+	}
+	case Kind::Tuple:
+	{
+		auto const& tupleType = dynamic_cast<TupleType const&>(type);
+		std::vector<std::string> memberNames;
+		std::vector<SortPointer> memberTypes;
+		for (auto&& [name, sortId] : tupleType.m_accessors)
+		{
+			memberNames.push_back(name);
+			memberTypes.push_back(unresolve(sortId));
+		}
+		return std::make_shared<TupleSort>(tupleType.m_name, std::move(memberNames), std::move(memberTypes));
+	}
+	default:
+		smtAssert(false, "Invalid SMT sort");
+	}
+}
+
+SortId SMTLib2Context::resolveBitVectorSort(BitVectorSort const& _sort)
+{
+	auto size = _sort.size;
+	auto it = m_bitVectorSorts.find(size);
+	if (it == m_bitVectorSorts.end())
+	{
+		auto newId = static_cast<uint32_t>(m_knownTypes.size());
+		m_knownTypes.emplace_back(std::make_unique<SMTLibSort>(Kind::BitVector, "(_ BitVec " + std::to_string(size) + ')', std::vector<SortId>{}, SortId{newId}));
+		auto&& [newIt, inserted] = m_bitVectorSorts.emplace(size, SortId{newId});
+		smtAssert(inserted);
+		it = newIt;
```

Comment:
This is perfectly fine like it is. But I'm a fan of immutability. I would prefer `return newIt->second` than assigning to `it`. Same for the next function. But your approach is fine too, do as you like more.

Reply:
Changed according to your suggestion.